### PR TITLE
Pose exp comp AKA discretization

### PR DESCRIPTION
**Pull request by greenblitz4590**

### CODE REVIEW

**Review by tnrz135 — APPROVED** [review]

File: src/main/java/edu/greenblitz/robotName/Robot.java [hunk 1]
```diff
@@ -28,7 +29,7 @@ public void robotInit() {
         initializeLogger();
 
         SwerveChassis.init();
-        SwerveChassis.getInstance().setDefaultCommand(new MoveByJoysticks(MoveByJoysticks.DriveMode.NORMAL));
+        SwerveChassis.getInstance().setDefaultCommand(new MoveByJoysticks(MoveByJoysticks.DriveMode.SLOW));
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/Robot.java:32]: if you merge change back to NORMAL

Comment (NoamRosenberg08) [reply src/main/java/edu/greenblitz/robotName/Robot.java:32]: for testing keep it slow

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java [hunk 2]
```diff
@@ -223,6 +229,10 @@ public void moveByChassisSpeeds(double forwardSpeed, double leftwardSpeed, doubl
                 angSpeed,
                 currentAng
         );
+        double timeStep = TIME_STEP;
+        if (RobotConstants.ROBOT_TYPE.equals(Robot.RobotType.ROBOT_NAME))
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java:233]: add {} for the if statement

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Modules/mk4iSwerveModule/MK4iSwerveConstants.java [hunk 3]
```diff
@@ -9,10 +9,10 @@
...
 	
-	public static SwerveModuleConfigObject MK4I_MODULE_FRONT_LEFT = new SwerveModuleConfigObject(1, 0, 1, Rotation2d.fromRotations(0.8486328125),false);
-	public static SwerveModuleConfigObject MK4I_MODULE_FRONT_RIGHT = new SwerveModuleConfigObject(3, 2, 2,Rotation2d.fromRotations(0.2939453125) ,true);
-	public static SwerveModuleConfigObject MK4I_MODULE_BACK_LEFT = new SwerveModuleConfigObject(5, 4, 3, Rotation2d.fromRotations(0.3024),false);
-	public static SwerveModuleConfigObject MK4I_MODULE_BAK_RIGHT = new SwerveModuleConfigObject(7, 6, 4, Rotation2d.fromRotations(0.8718),true);
+	public static SwerveModuleConfigObject MK4I_MODULE_FRONT_LEFT = new SwerveModuleConfigObject(1, 10,0, Rotation2d.fromRotations(0.8486328125),false);
+	public static SwerveModuleConfigObject MK4I_MODULE_FRONT_RIGHT = new SwerveModuleConfigObject(3, 11, 2,Rotation2d.fromRotations(0.2939453125) ,true);
+	public static SwerveModuleConfigObject MK4I_MODULE_BACK_LEFT = new SwerveModuleConfigObject(2, 8, 1, Rotation2d.fromRotations(0.3024),false);
+	public static SwerveModuleConfigObject MK4I_MODULE_BAK_RIGHT = new SwerveModuleConfigObject(12, 5, 3, Rotation2d.fromRotations(0.8718),true);
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/Modules/mk4iSwerveModule/MK4iSwerveConstants.java:15]: why did you change the values? you have an MK4I swerve chassis in your basement?

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/SwerveChassisUtils.java [hunk 4]
```diff
@@ -6,22 +6,11 @@ public class SwerveChassisUtils {
...
-        double factoredOutputValue;
-        if (Math.abs(joystickValue * velocityFactor) < velocityFactor) {
-            factoredOutputValue = joystickValue * velocityFactor;
-        } else {
-            factoredOutputValue = velocityFactor;
-        }
-        return factoredOutputValue * (ChassisConstants.ANGULAR_JOYSTICK_INVERTED ? -1 : 1);
+        double factoredOutputValue = joystickValue * velocityFactor;
+        return Math.min(factoredOutputValue, velocityFactor) * (ChassisConstants.ANGULAR_JOYSTICK_INVERTED ? -1 : 1);
     }
```
Comment (NoamRosenberg08) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/SwerveChassisUtils.java:15]: for another pull request. please remove when you merge.

**Review by NoamRosenberg08 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java [hunk 5]
```diff
@@ -223,16 +231,31 @@ public void moveByChassisSpeeds(double forwardSpeed, double leftwardSpeed, doubl
...
+        double timeStep = TIME_STEP*SIMULATION_DISCRETION_CONSTANT;
+        if (RobotConstants.ROBOT_TYPE.equals(Robot.RobotType.ROBOT_NAME)) {
+            timeStep = RoborioUtils.getCurrentRoborioCycle()*REAL_DISCRETION_CONSTANT;
+        }
+        chassisSpeeds = ChassisSpeeds.discretize(chassisSpeeds,timeStep);
         SwerveModuleState[] states = kinematics.toSwerveModuleStates(chassisSpeeds);
         SwerveModuleState[] desaturatedStates = desaturateSwerveModuleStates(states);
         setModuleStates(desaturatedStates);
     }
+    public static void printStates(SwerveModuleState[] states) {
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java:243]: why static

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java [hunk 6]
```diff
@@ -223,16 +231,31 @@ public void moveByChassisSpeeds(double forwardSpeed, double leftwardSpeed, doubl
...
+        for (int i = 0; i < states.length; i++) {
+            System.out.println(states[i].angle + "," + states[i].speedMetersPerSecond);
```
Comment (bossyshrimp68) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java:245]:
```suggestion
            System.out.println(states[i].angle + ", " + states[i].speedMetersPerSecond);
```

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]
amazing and breathtaking (the ", " comment is nazi you dont really need to change)

**Review by bossyshrimp68 — APPROVED** [review]
Romy - good job yoni gg

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java [hunk 7]
```diff
@@ -223,16 +231,30 @@ public void moveByChassisSpeeds(double forwardSpeed, double leftwardSpeed, doubl
...
         setModuleStates(desaturatedStates);
     }
+    private double getDiscretizedTimeStep() {
+        double timeStep = TIME_STEP*SIMULATION_DISCRETION_CONSTANT;
+        if (RobotConstants.ROBOT_TYPE.equals(Robot.RobotType.ROBOT_NAME)) {
+            timeStep = RoborioUtils.getCurrentRoborioCycle()*REAL_DISCRETION_CONSTANT;
+        }
+        return timeStep;
+    }
+    public void printStates(SwerveModuleState[] states) {
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java:246]: remove debug func

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java [hunk 8]
```diff
@@ -223,16 +231,30 @@ public void moveByChassisSpeeds(double forwardSpeed, double leftwardSpeed, doubl
...
         );
+        chassisSpeeds = ChassisSpeeds.discretize(chassisSpeeds, getDiscretizedTimeStep());
         SwerveModuleState[] states = kinematics.toSwerveModuleStates(chassisSpeeds);
         SwerveModuleState[] desaturatedStates = desaturateSwerveModuleStates(states);
         setModuleStates(desaturatedStates);
     }
+    private double getDiscretizedTimeStep() {
+        double timeStep = TIME_STEP*SIMULATION_DISCRETION_CONSTANT;
+        if (RobotConstants.ROBOT_TYPE.equals(Robot.RobotType.ROBOT_NAME)) {
+            timeStep = RoborioUtils.getCurrentRoborioCycle()*REAL_DISCRETION_CONSTANT;
```
Comment (yonnbro) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java:242]: CTRL + ALT + L

**Review by yonnbro — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java (hunk 7 above)
Comment (yonnbro) [reply src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java:246]: or use it in a dashboard

**Review by bossyshrimp68 — APPROVED** [review]

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/ChassisConstants.java [hunk 9]
```diff
@@ -32,6 +35,10 @@ public class ChassisConstants {
     public static final double DRIVER_LINEAR_SPEED_FACTOR_SLOW = 0.5;
 
     public static final boolean ANGULAR_JOYSTICK_INVERTED = true;
+
+    public static final boolean LINEAR_JOYSTICK_INVERTED = false;
+    public static final MoveByJoysticks.DriveMode CURRENT_DRIVE_MODE = MoveByJoysticks.DriveMode.NORMAL;
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/ChassisConstants.java:40]: Why is `CURRENT_DRIVE_MODE` final?

The name implies that this value changes throughout the runtime

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java [hunk 10]
```diff
@@ -27,6 +32,9 @@
 
 import java.util.Optional;
 
+import static edu.greenblitz.robotName.RobotConstants.SimulationConstants.TIME_STEP;
+import static edu.greenblitz.robotName.subsystems.swerve.Chassis.ChassisConstants.*;
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java:36]: It is considered a bad practice to `import *`, make sure you really need it

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java [hunk 11]
```diff
@@ -213,11 +221,26 @@ public Rotation2d getChassisAngle() {
...
 		}
 	}
+  	private double getDiscretizedTimeStep() {
+    	double timeStep = getActualTimeStep();
+    	double discretizedTimeStep = timeStep * FAST_DISCRETION_CONSTANT;
+    	if (CURRENT_DRIVE_MODE.equals(MoveByJoysticks.DriveMode.SLOW)) {
+      		discretizedTimeStep = timeStep * SLOW_DISCRETION_CONSTANT;
+    	}
+    	return discretizedTimeStep;
+  	}
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java:236]: CTRL + ALT + L

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java [hunk 12]
```diff
@@ -213,11 +221,26 @@ public Rotation2d getChassisAngle() {
...
 		for (Module module : Module.values()) {
 			setModuleStateForModule(module, states[module.ordinal()]);
 		}
 	}
+  	private double getDiscretizedTimeStep() {
+    	double timeStep = getActualTimeStep();
+    	double discretizedTimeStep = timeStep * FAST_DISCRETION_CONSTANT;
+    	if (CURRENT_DRIVE_MODE.equals(MoveByJoysticks.DriveMode.SLOW)) {
+      		discretizedTimeStep = timeStep * SLOW_DISCRETION_CONSTANT;
+    	}
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java:234]: This is a bit of a weird way to handle an enum

File: src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java [hunk 13]
```diff
@@ -213,11 +221,26 @@ public Rotation2d getChassisAngle() {
...
+    	return discretizedTimeStep;
+  	}
+  	private double getActualTimeStep() {
+    	if (RobotConstants.ROBOT_TYPE.equals(Robot.RobotType.ROBOT_NAME))
+      		return RoborioUtils.getCurrentRoborioCycle();
+		return TIME_STEP;
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/robotName/subsystems/swerve/Chassis/SwerveChassis.java:240]: Bruh

**Review by Stormageddon37 — CHANGES_REQUESTED** [review]

**Review by bossyshrimp68 — APPROVED** [review]

**Pull request merged**